gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1426 @@
+            }
+          }
+        }
+        if (rrset.getType().equals("SOA") && findByNameAndType(deletions, null, "SOA") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.additions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'SOA' at the apex.", index));
+        }
+        if (rrset.getType().equals("NS") && findByNameAndType(deletions, null, "NS") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.additions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'NS' at the apex.", index));
+        }
+        index++;
+      }
+    }
+    if (deletions != null) {
+      int index = 0;
+      for (ResourceRecordSet rrset : deletions) {
+        if (rrset.getType().equals("SOA") && findByNameAndType(additions, null, "SOA") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.deletions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'SOA' at the apex.", index));
+        }
+        if (rrset.getType().equals("NS") && findByNameAndType(additions, null, "NS") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.deletions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'NS' at the apex.", index));
+        }
+        index++;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Helper for searching rrsets in a collection.
+   */
+  private static ResourceRecordSet findByNameAndType(Iterable<ResourceRecordSet> records,
+      String name, String type) {
+    if (records != null) {
+      for (ResourceRecordSet rrset : records) {
+        if ((name == null || name.equals(rrset.getName()))
+            && (type == null || type.equals(rrset.getType()))) {
+          return rrset;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * We only provide the most basic validation for A and AAAA records.
+   */
+  static boolean checkRrData(String data, String type) {
+    // todo add validation for other records
+    String[] tokens;
+    switch (type) {
+      case "A":
+        tokens = data.split("\\.");
+        if (tokens.length != 4) {
+          return false;
+        }
+        for (String token : tokens) {
+          try {
+            Integer number = Integer.valueOf(token);
+            if (number < 0 || number > 255) {
+              return false;
+            }
+          } catch (NumberFormatException ex) {
+            return false;
+          }
+        }
+        return true;
+      case "AAAA":
+        tokens = data.split(":", -1);
+        if (tokens.length != 8) {
```

Comment:
Talking about  2607:f8b0:400a:801::1005. The two colons one after another can expand to a sequence of :(0:)*: so it is valid.